cooked-validators/src/Cooked/Attack/Tweak/AddInputsAndOutputs.hs

```diff
@@ -0,0 +1,98 @@
+{-# HLINT ignore "Use <&>" #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+-- | Some 'Tweak's that add or remove inputs and outputs from transactions.
+module Cooked.Attack.Tweak.AddInputsAndOutputs where
+
+import Control.Monad
+import Cooked.Attack.Tweak.Common
+import Cooked.Tx.Constraints.Type
+import Data.List
+import qualified Data.Map as Map
+import qualified Ledger as Pl
+import Optics.Core
+import Test.QuickCheck (NonZero (..))
+
+-- * Adding and removing transaction inputs
+
+-- | Ensure that a given 'SpendableOut' is being spent with a given
+-- 'TxSkelIn'. The return value will be @Just@ the added data, if anything
+-- changed.
+ensureInputTweak :: SpendableOut -> TxSkelIn -> Tweak (Maybe (SpendableOut, TxSkelIn))
+ensureInputTweak sOut howConsumed = do
+  presentInputs <- viewTweak txSkelInsL
+  case presentInputs Map.!? sOut of
+    Nothing -> doTheChange
+    Just howConsumed' ->
+      if howConsumed == howConsumed'
+        then return Nothing
+        else doTheChange
```

Comment:
I think you can do
```suggestion
  if presentInputs Map.!? sOut == Just howConsumed
    then return Nothing
    else doTheChange
```
(perhaps inlining `doTheChange` now that it's only used once).